### PR TITLE
add committed index in Progress

### LINKED ISSUE #325: Add a way to figure out followers commit index
Raft doesn't requires follower reports its commit index. But aggregated them can help us find out whether a log is committed in all nodes, which is quite useful in some cases.

For example, in TiKV, when enabling hibernate regions, it's better to let all the followers to commit to latest logs before going to sleep, otherwise follower can wait for a long time before committing and applying logs, which can hurt follower reads.

It can also be used in merge too.

### PR DESCRIPTION
**Pull request by accelsao**
Close #325

- [X] Commit index won't step backward.
- [X] Log rejection should not change commit index
- [X] Successfully Append should update commit index
- [x] Heartbeat should update commit index

Signed-off-by: accelsao <bebe8277@gmail.com>

### CODE REVIEW

**Comment by accelsao** [comment]
I think my test_case is useless..

**Comment by accelsao** [comment]
error `The command "if [[ $TRAVIS_RUST_VERSION == "stable" && $TRAVIS_OS_NAME == "linux" ]]; then cargo fmt --all -- --check; fi" exited with 1.` in travis build is left

File: harness/tests/integration_cases/test_raft.rs [hunk 1]
```diff
@@ -212,6 +212,121 @@ fn test_progress_update() {
     }
 }
 
+#[test]
+fn test_progress_committed() {
```
Comment (BusyJay) [inline harness/tests/integration_cases/test_raft.rs:216]: I think you can use `Network` to test the commit index change. It should validate following case:
1. Log rejection should not change commit index
2. Successfully Append should update commit index
3. Heartbeat should update commit index
4. Commit index won't step backward.

Comment (accelsao) [reply harness/tests/integration_cases/test_raft.rs:216]: I still do not know when will we get the smaller commit index

Comment (BusyJay) [reply harness/tests/integration_cases/test_raft.rs:216]: For example, leader sends two `MsgAppend` to follower and gets `MsgAppendResponse` in reverse order.

File: harness/tests/integration_cases/test_raft.rs [hunk 2]
```diff
@@ -212,6 +211,350 @@ fn test_progress_update() {
...
+        4
+    );
+
+    nt.isolate(2);
+    let m = new_message_with_entries(2, 2, MessageType::MsgPropose, vec![test_entries]);
+    nt.send(vec![m.clone(), m]);
+    nt.recover();
+
+    // test append entries rejection (update committed index = false)
+    let mut m = new_message(2, 1, MessageType::MsgAppend, 0);
```
Comment (BusyJay) [inline harness/tests/integration_cases/test_raft.rs:385]: Can `dispatch` and `read_messages` be used to avoid creating messages?

Comment (accelsao) [reply harness/tests/integration_cases/test_raft.rs]: thank advice, it makes life easier

File: harness/tests/integration_cases/test_raft.rs [hunk 3]
```diff
@@ -212,6 +211,166 @@ fn test_progress_update() {
...
+
+    // #4 pass a smaller committed index, it occurs when the append response delay
+    let mut msg_append_response = new_message(2, 1, MessageType::MsgAppendResponse, 0);
+    msg_append_response.term = 3;
+    msg_append_response.index = 10;
+    msg_append_response.commit = 10;
+
+    nt.send(vec![msg_append_response.clone()]);
+
+    assert_eq!(nt.peers[&1].prs().get(2).unwrap().committed_index, 10);
```
Comment (BusyJay) [inline harness/tests/integration_cases/test_raft.rs:364]: This seems a bug.

Comment (accelsao) [reply harness/tests/integration_cases/test_raft.rs:364]: shouldnt that commiitted index is update as 10?

Comment (BusyJay) [reply harness/tests/integration_cases/test_raft.rs]: If leader doesn't have log 10, it shouldn't commit to 10. Giving that raft-rs doesn't take much effort in byzantine fault tolerant, I think remove the case is OK for now.

File: harness/tests/integration_cases/test_raft.rs [hunk 4]
```diff
@@ -212,6 +211,192 @@ fn test_progress_update() {
...
+            nt.peers[&1].prs().get(1).unwrap().committed_index,
+            nt.peers[&1].prs().get(2).unwrap().committed_index,
+            nt.peers[&1].prs().get(3).unwrap().committed_index
+        ),
+        (0, 3, 1)
+    );
+
+    // #2 test heart beat
+    let heart_beat = new_message(1, 1, MessageType::MsgBeat, 0);
+    nt.send(vec![heart_beat]);
```
Comment (gengliqi) [inline harness/tests/integration_cases/test_raft.rs:260]: `heartbeat` is a word.

File: src/raft.rs [hunk 5]
```diff
@@ -1939,6 +1945,7 @@ impl<T: Storage> Raft<T> {
             .maybe_append(m.index, m.log_term, m.commit, &m.entries)
         {
             to_send.set_index(last_idx);
+            to_send.set_commit(self.raft_log.committed);
             self.send(to_send);
```
Comment (gengliqi) [inline src/raft.rs:1949]: How about moving these two lines and 1964,1965 lines outside this conditional branch?

**Pull request merged**